### PR TITLE
CNV-19687: Sidebar editor

### PR DESCRIPTION
**Pull request by upalatucci**
## 📝 Description

Form - yaml editor side by side.
The user should be able to see how the changes in the forms affect the yaml

We have here 3 new components.

Context, Switch, and Editor

Follow-up pr would add this feature also to the wizard and to the templates

## 🎥 Demo

Screenshot from 2022-09-02 15-31-26
Screenshot from 2022-09-02 15-31-20
Screenshot from 2022-09-02 15-31-13

### CODE REVIEW

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+const SidebarEditor: React.FC<SidebarEditorProps> = ({ children, resource }) => {
+  const { showEditor } = React.useContext(SidebarEditorContext);
+  return (
+    <Sidebar isPanelRight hasGutter hasNoBackground className="sidebar-editor">
+      <SidebarContent>{children}</SidebarContent>
+      {showEditor && (
+        <SidebarPanel
+          width={{ default: 'width_33', lg: 'width_50', xl: 'width_50' }}
+          style={{ minHeight: '100%' }}
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:29]: cant be in the classname ?

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+          className="sidebar-editor"
+        >
+          <React.Suspense fallback={<Loading />}>
+            <YAMLEditor value={dump(resource)} minHeight="300px" />
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:33]: dump may result in exception, should be called outside with try catch block

File: src/utils/components/SidebarEditor/SidebarEditorContext.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+
+export type SidebarEditorContextType = {
+  showEditor: boolean;
+  setEditorVisible?: (editorVisible: boolean) => void;
+};
+
+export const SidebarEditorContext = React.createContext<SidebarEditorContextType>({
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditorContext.tsx:8]: why do we need context for this component?

**Review by metalice — COMMENTED** [review]
a few months ago we updated tsconfig, so we can import react normally and not * as react , so no need anymore for Reac.useMemo or React.useCallback, you can simply import useMemo or useCallback

**Comment by metalice** [comment]
is the editor it not editable? what happens if a user change values in YAML? will it reflect in the form?

**Comment by upalatucci** [comment]
The editor is editable, unfortunately. I would like not to be like that. As it is the values would not be reflected in the form.

**Comment by upalatucci** [comment]
aaah but there is a readonly option that I can use!

File: src/utils/components/SidebarEditor/SidebarEditor.tsx (hunk 2 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:33]: we'll use dump with `skipInvalid` option

File: src/utils/components/SidebarEditor/utils.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { load } from 'js-yaml';
+
+export const safeLoad = <T>(value: string): T | undefined => {
```
Comment (metalice) [inline src/utils/components/SidebarEditor/utils.ts:3]: what about safeDump?
if it's a generic type, I don't think nested util file under local utils is the correct place, as here it is always Resource type

File: src/utils/components/SidebarEditor/sidebar-editor.scss [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  .pf-c-sidebar__main {
+    min-height: 100%;
+  }
+
+  .pf-c-sidebar__main {
+    align-items: stretch;
+  }
+
+  .pf-c-sidebar__panel {
+    padding: 4px;
```
Comment (metalice) [inline src/utils/components/SidebarEditor/sidebar-editor.scss:14]: can be changed to a var?

File: src/utils/components/SidebarEditor/SidebarEditorSwitch.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  const { t } = useKubevirtTranslation();
+  const { showEditor, setEditorVisible } = React.useContext(SidebarEditorContext);
+
+  return (
+    <Switch
+      id="sidebar-editor-switch"
+      label={t('YAML: ON')}
+      labelOff={t('YAML: OFF')}
+      isChecked={showEditor}
+      onChange={(checked) => setEditorVisible(checked)}
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditorSwitch.tsx:18]: onChange={setEditorVisible}

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    reloadResource();
+  }, [reloadResource]);
+
+  const changeResource = (newValue: string) => {
+    setEditableYAML(newValue);
+
+    return {};
+  };
+
+  const edited = editableYAML !== resourceYAML;
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:54]: will calculate on each render.. also if used only once you can simply use it at line 76

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  resource,
+  onResourceUpdate,
+}: SidebarEditorProps<Resource>): JSX.Element => {
+  const resourceYAML = React.useMemo(() => dump(resource, { skipInvalid: true }), [resource]);
+  const { showEditor } = useContext(SidebarEditorContext);
+  const [editableYAML, setEditableYAML] = useState('');
+  const editedResource = safeLoad<Resource>(editableYAML);
+
+  const reloadResource = useCallback(() => {
+    setEditableYAML(resourceYAML);
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:41]: can be added to line 35

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  }, [resourceYAML]);
+
+  useEffect(() => {
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:44]: can be removed if reloadResource callback is added at line 35

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    reloadResource();
+  }, [reloadResource]);
+
+  const changeResource = (newValue: string) => {
+    setEditableYAML(newValue);
+
+    return {};
+  };
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:52]: no need, can simply declare at line 67

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+      <SidebarContent>{children && children(editedResource ?? resource)}</SidebarContent>
+      {showEditor && (
+        <SidebarPanel
+          width={{ default: 'width_33', lg: 'width_50', xl: 'width_50' }}
+          className="sidebar-editor"
+        >
+          <Stack hasGutter>
+            <StackItem isFilled>
+              <React.Suspense fallback={<Loading />}>
+                <YAMLEditor value={editableYAML} minHeight="300px" onChange={changeResource} />
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:67]: onChange={setEditableYAML}

Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:67]: can we set min height via classname?

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    setEditableYAML(newValue);
+
+    return {};
+  };
+
+  const edited = editableYAML !== resourceYAML;
+
+  return (
+    <Sidebar isPanelRight hasGutter hasNoBackground className="sidebar-editor">
+      <SidebarContent>{children && children(editedResource ?? resource)}</SidebarContent>
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:58]: there is no check for children function?

Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:58]: Check is in the type props. Typescript will raise an error if SidebarEditor will be used wrongly

File: src/utils/components/SidebarEditor/SidebarEditor.tsx (hunk 11 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:67]: unfortunately, the yaml editor has wrong props. onChange and onSave want an object as return value that in the code is not used, for this reason i return an empty object.
I have a pr for this.

File: src/utils/components/SidebarEditor/SidebarEditor.tsx (hunk 10 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:52]: same https://github.com/kubevirt-ui/kubevirt-plugin/pull/845#discussion_r968098081

File: src/utils/components/SidebarEditor/SidebarEditor.tsx (hunk 7 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:54]: okay. Removed. the save button will be always enabled. Not too bad

File: src/utils/components/SidebarEditor/SidebarEditor.tsx (hunk 11 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:67]: no className props. minHeight is used for inline style in the source code combined with react-measure. But I can do it with the parent

File: src/utils/components/SidebarEditor/utils.ts (hunk 4 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/utils.ts:3]: the doc about `skipInvalid` says `do not throw on invalid types (like function in the safe schema) and skip pairs and single values with such types.`. Can we trust it?

File: src/utils/components/SidebarEditor/SidebarEditor.tsx (hunk 8 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditor.tsx:41]: good catch

File: src/utils/components/SidebarEditor/utils.ts (hunk 4 above)
Comment (upalatucci) [reply src/utils/components/SidebarEditor/utils.ts:3]: `Resource` in SidebarEditor is a generic type. Can be VM, VMI, Template, etc.

File: src/utils/components/SidebarEditor/SidebarEditor.tsx [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+  children: (resource: Resource) => React.ReactNode;
+};
+
+const SidebarEditor = <Resource extends K8sResourceCommon>({
+  children,
+  resource,
+  onResourceUpdate,
+}: SidebarEditorProps<Resource>): JSX.Element => {
+  const [editableYAML, setEditableYAML] = useState('');
+  const resourceYAML = React.useMemo(() => {
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditor.tsx:36]: can use simply useMemo

File: src/utils/components/SidebarEditor/SidebarEditorContext.tsx [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  showEditor: boolean;
+  setEditorVisible?: (editorVisible: boolean) => void;
+};
+
+export const SidebarEditorContext = createContext<SidebarEditorContextType>({
+  showEditor: false,
+});
+
+export const SidebarEditorProvider: FC = ({ children }) => {
+  const [showEditor, setEditor] = useState(false);
```
Comment (metalice) [inline src/utils/components/SidebarEditor/SidebarEditorContext.tsx:13]: usually the set funciton is exact name of the first variable, setShowEditor , also, do u really need the useCallback on line 15? i think you can simply transfer the setShowEditor itself

Comment (upalatucci) [reply src/utils/components/SidebarEditor/SidebarEditorContext.tsx]: Fixed

**Pull request merged**